test/e2e_node/e2e_service.go

```diff
@@ -33,22 +34,181 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"github.com/kardianos/osext"
 
 	"k8s.io/kubernetes/test/e2e/framework"
 )
 
+// TODO(random-liu): Move this file to a separate package.
 var serverStartTimeout = flag.Duration("server-start-timeout", time.Second*120, "Time to wait for each server to become healthy.")
 
+// E2EServices starts and stops e2e services. The test use it to start and
+// stop all dependent e2e services.
+type E2EServices struct {
+	cmd *exec.Cmd
+}
+
+func NewE2eServices() *E2EServices {
+	return &E2EServices{}
+}
+
+// StartE2eServices start the e2e services in another process, it returns
+// when all e2e services are ready.
+// We want to statically link e2e services into the test binary, but we don't
+// want their glog to pollute the test result. So we run the binary in start-
+// services-only mode to start e2e services in another process.
+func (e *E2EServices) StartE2EServices() (string, error) {
+	// Create the manifest path for kubelet.
+	// TODO(random-liu): Remove related logic when we move kubelet starting logic out of the test.
+	manifestPath, err := ioutil.TempDir("", "node-e2e-pod")
+	if err != nil {
+		return "", err
+	}
+	testBin, err := osext.Executable()
+	if err != nil {
+		return manifestPath, fmt.Errorf("can't get current binary: %v", err)
+	}
+	cmd := exec.Command(testBin,
+		"--start-services-only",
```

Comment:
I think I get what you are trying to do. Instead how about this:

1. Invoke `e2e-node.test`
1. `e2e-node.test` will internally exec `e2e-node.test apiserver` & `e2e-node.test etcd`
1. `e2e-node.test` will run the tests only the exec'ed processes are running.
1. `e2e-node.test` will evaluate os.Argv[1] to determine if it has to start etcd or apiserver, or run in test mode.

Also consider using env variables instead of flags if you'd like to hide some of the implementation details.